src/Voron/Impl/LowLevelTransaction.cs

```diff
@@ -934,6 +935,9 @@ public void Dispose()
             if (_txState.HasFlag(TxState.Disposed))
                 return;
 
+            if (Flags == TransactionFlags.ReadWrite && NativeMemory.CurrentThreadStats != CurrentTransactionHolder)
+                ThrowDisposeOfTxMustBeCalledOnTheSameThreadThatCreatedIt();
```

Comment:
That _didn't_ used to be something that we require.
But I can't think of a valid scenario now that we don't actually do that.

Reply:
Indeed there isn't any usage like according to the tests. 

I have tried different approaches - a lock between the commit and the dispose or additional flags to mark that we're committing and disposing but these had other issues (mostly because the commit is a multi phase process - especially for transactions with async commit usage).

This check was the simplest solution because it takes usage of the already existing lock for write transactions to detect the issue - `SemaphoreSlim _transactionWriter`. But we cannot check directly whether we're the one who acquired the lock so we do it via `_currentWriteTransactionHolder` the is assigned to `tx.CurrentTransactionHolder` for write transactions:

https://github.com/ravendb/ravendb/blob/e122a905a773a9e2f16f4458bc953709ee3f789f/src/Voron/StorageEnvironment.cs#L727-L729

Reply:
### Issue with async code

This will not work with async code:

```
using (context.OpenWriteTransaction())
{
    await ...;
   
    context.Transaction.Commit();
}
```

After `await` we _might_ be no a different thread. So thread based check that I added here will fail. 

### Advantages 

But it might be a good thing because we don't want to have `await ...` under the write transaction lock. It is a potentially  long running operation while we're holding the write tx lock. So this check gives a chance to detect that.

### Risk

At the same time it's very risky because if we throw here, while we were really the one who opened the write tx, **then we will never release the tx lock**. That would leave the storage environment in unusable state.

---

The issue was found during execution of `RachisTests.BasicTests.RavenDB_13659` test which has async code involved in the scope of the write transaction:

https://github.com/ravendb/ravendb/blob/9684dcbebd0272d0322a7c792e28ec5e9a55545d/test/RachisTests/BasicTests.cs#L91-L96

Note that it's not deterministic - sometimes it might fail while sometimes it won't (e.g. it didn't fail during execution of test in this PR - neither in Release nor Debug).

Searching for similar usage in RavenDB code resulted in finding the following piece of code which also has `await` during execution of write transaction:

https://github.com/ravendb/ravendb/blob/9684dcbebd0272d0322a7c792e28ec5e9a55545d/src/Raven.Server/Documents/ReplayTxCommandHelper.cs#L47-L63

---

### Proposed approach

There might be more tricky places where we have `await` inside write transactions and that won't be easily visible. It's very challenging to identify all of them. There is a concern that this check might affect a production system.

Let's have this check _only_ in Debug more and fix the places that we're aware of.

Reply:
Note that this is not a valid scenario as far as we are concerned.
I'm fine with making this in debug only, because under no circumstances should we allow it ever.

Reply:
I concur, I have been adding myself checks all over the place in Debug mode. I think it is a good idea to have extra checks on clearly erroneous scenarios in debug mode.